sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -2767,7 +2767,12 @@ class Dataset[T] private[sql](
    * @since 1.6.0
    */
   def count(): Long = withAction("count", groupBy().count().queryExecution) { plan =>
-    plan.executeCollect().head.getLong(0)
+    val collected = plan.executeCollect()
+    if (collected.isEmpty) {
+      0
+    } else {
+      collected.head.getLong(0)
+    }
```

Comment:
`spark.range(-10, -9, -20, 1).select("id").count` in `DataFrameRangeSuite` causes exception here. `plan.executeCollect().head` pulls empty iterator by calling `next`.

Reply:
I think it is caused by returning `SinglePartition` when there is no data (and therefore no partition). So I think we should fix it there and not here.

Reply:
Right, making sense.